crates/turborepo-lib/src/shim/local_turbo_config.rs

```diff
@@ -0,0 +1,82 @@
+use jsonc_parser::{
+    ast::{ObjectPropName, Value},
+    parse_to_ast,
+};
+use turborepo_repository::{inference::RepoState, package_manager::PackageManager};
+
+/// Struct containing information about the desired local turbo version
+/// according to lockfiles, package.jsons, and if all else fails turbo.json
+pub struct LocalTurboConfig {
+    turbo_version: String,
+}
+
+impl LocalTurboConfig {
+    pub fn infer(repo_state: &RepoState) -> Option<Self> {
+        let turbo_version = Self::turbo_version_from_lockfile(repo_state)
+            .or_else(|| Self::turbo_version_from_package_json(repo_state))
+            .or_else(|| Self::turbo_version_from_turbo_json_schema(repo_state))?;
+        Some(Self { turbo_version })
+    }
+
+    pub fn turbo_version(&self) -> &str {
+        &self.turbo_version
+    }
+
+    fn turbo_version_from_lockfile(repo_state: &RepoState) -> Option<String> {
+        if let Ok(package_manager) = &repo_state.package_manager {
+            let lockfile = package_manager
+                .read_lockfile(&repo_state.root, &repo_state.root_package_json)
+                .ok()?;
+            return lockfile.turbo_version();
+        }
+
+        // If there isn't a package manager, just try to parse all known lockfiles
+        // This isn't the most effecient, but since we'll be hitting network to download
+        // the correct binary the unnecessary file reads.
+        PackageManager::supported_managers().iter().find_map(|pm| {
+            let lockfile = pm
+                .read_lockfile(&repo_state.root, &repo_state.root_package_json)
+                .ok()?;
```

Comment:
If parsing the back-up fails perhaps we just continue and try others rather than exit? Your call, thought I'd mention it.

Reply:
We still do since the `?` is in the `find_map` closure it'll just short circuit the current package manager check and move onto the next one.